experiment_code/hackrl/experiment.py

```diff
@@ -139,6 +140,61 @@ def _iter():
                     )
                     list(self.threadpool.map(convert, range(self.ttyrec_batch_size)))
 
+                    # Flatten and convert to numpy array
+                    tty_chars = (
+                        mb_tensors["tty_chars"][..., -2:, :].flatten(start_dim=2).detach().cpu().numpy()
+                    )
+
+                    # Convert numeric values to characters
+                    characters = np.apply_along_axis(
+                        lambda x: "".join(map(chr, x)), axis=2, arr=tty_chars
+                    )
+
+                    # Split characters by space
+                    words = np.char.split(characters, " ")
+
+                    # Extract Dlvl values from words
+                    dlvl = np.array(
+                        [
+                            [
+                                [w.split(":")[-1] for w in string if w.startswith("Dlvl:")]
+                                for string in sublist
+                            ]
+                            for sublist in words
+                        ]
+                    )
+
+                    def _filter_2d_dlvl_array(frame):
+                        def filter_element(x):
+                            if isinstance(x, list) and len(x) > 0:
+                                return int(x[0])
+                            else:
+                                return None
+
+                        # Vectorize the filtering function
+                        filter_vectorized = np.vectorize(filter_element, otypes=[object])
+
+                        # Apply the filtering function to the frame
+                        filtered_frame = filter_vectorized(frame)
+
+                        # Replace None values with element on the right or the left
```

Comment:
Since you are using a for loop you could just remember the previous dlvl and use its value to fill the Nones. It will be much simpler. You cannot go to the next level with a popup window.

Reply:
```
                        # Replace None values with dlvl from previous frame
                        for i in range(filtered_frame.shape[0]):
                            prev_dlvl = 1
                            for j in range(filtered_frame.shape[1]):
                                if filtered_frame[i, j] is None:
                                    filtered_frame[i, j] = prev_dlvl
                                prev_dlvl = filtered_frame[i, j]
```

Reply:
Actually, the code above is also wrong. It doesn't handle the case when `filtered_frame[i, 0] is None and filtered_frame[i, j] > 1`. Your version also would be able to handle it. It's possible to pass the prev_dlvl as a state, but I think it's better to just use the future or past dlvl. So your idea is fine with a modification to look further (it's possible that there will be more Nones than 2.

Reply:
Actually, never mind. prev_dlvl as a state is easy to implement and I've done it. You can see a new commit.